crates/services/txpool_v2/src/storage/mod.rs

```diff
@@ -36,34 +38,53 @@ pub struct StorageData {
 
 pub type RemovedTransactions = Vec<PoolTransaction>;
 
+pub trait Collision<StorageIndex> {
+    /// Returns the instigator of the collision.
+    fn tx(&self) -> &PoolTransaction;
+
+    /// Returns the transactions that collide with the given transaction.
+    fn colliding_transactions(&self) -> &HashMap<StorageIndex, Vec<CollisionReason>>;
+
+    /// Unwraps the collision into a instigator transaction.
+    fn into_instigator(self) -> PoolTransaction;
+}
+
+pub trait CheckedCollision<StorageIndex>: Collision<StorageIndex> {
+    /// Returns the list of all dependencies of the transaction.
+    fn all_dependencies(&self) -> &HashSet<StorageIndex>;
+}
+
 /// The storage trait is responsible for storing transactions.
 /// It has to manage somehow the dependencies between transactions.
 pub trait Storage {
     /// The index type used in the storage and allow other components to reference transactions.
-    type StorageIndex: Copy + Debug;
+    type StorageIndex: Copy + Debug + Eq + Hash;
+    type CheckedCollision<C>: CheckedCollision<Self::StorageIndex>
+    where
+        C: Collision<Self::StorageIndex>;
 
     /// Store a transaction in the storage according to the dependencies.
     /// Returns the index of the stored transaction.
-    fn store_transaction(
+    fn store_transaction<C>(
         &mut self,
-        transaction: PoolTransaction,
+        checked_collision: Self::CheckedCollision<C>,
```

Comment:
That's soooo strange to me to have the `store_transaction` not have a `transaction` as parameter. is it possible to get explicitely wht's used inside this function: tx and dependencies ?

Reply:
The main idea is that `store_transaction` accepts the type returned by the `can_store_transaction`. From that perspective it is logical=)